synapseclient/models/factory_operations.py

```diff
@@ -0,0 +1,1209 @@
+
+    Arguments:
+        synapse_id: The Synapse ID of the entity to retrieve (e.g., 'syn123456').
+            Mutually exclusive with entity_name.
+        entity_name: The name of the entity to find. Must be used with this approach
+            instead of synapse_id. When looking up projects, parent_id should be None.
+        parent_id: The parent entity ID when looking up by name. Set to None when
+            looking up projects by name. Only used with entity_name.
+        version_number: The specific version number of the entity to retrieve. Only
+            applies to versionable entities (File, Table, Dataset). If not specified,
+            the most recent version will be retrieved. Ignored for other entity types.
+        activity_options: Activity-specific configuration options. Can be applied to
+            any entity type to include activity information.
+        file_options: File-specific configuration options. Only applies to File entities.
+            Ignored for other entity types.
+        table_options: Table-specific configuration options. Only applies to Table-like
+            entities (Table, Dataset, EntityView, MaterializedView, SubmissionView,
+            VirtualTable, DatasetCollection). Ignored for other entity types.
+        link_options: Link-specific configuration options. Only applies when the entity
+            is a Link. Ignored for other entity types.
+        synapse_client: If not passed in and caching was not disabled by
+            `Synapse.allow_client_caching(False)` this will use the last created
+            instance from the Synapse class constructor.
+
+    Returns:
+        The appropriate Synapse entity model instance based on the entity type.
+
+    Raises:
+        ValueError: If both synapse_id and entity_name are provided, or if neither is provided.
+        ValueError: If entity_name is provided without this being a valid lookup scenario.
+        ValueError: If the synapse_id is not a valid Synapse ID format.
+
+    Note:
+        When using entity_name lookup:
+
+        - For projects: leave parent_id=None
+        - For all other entities: provide the parent_id of the containing folder/project
+
+    Example: Retrieving entities by ID
+        Get any entity by Synapse ID:
+
+        ```python
+        from synapseclient import Synapse
+        from synapseclient.models import get
+
+        syn = Synapse()
+        syn.login()
+
+        # Works for any entity type
+        entity = get(synapse_id="syn123456")
+
+        # The returned object will be the appropriate type
+        if isinstance(entity, File):
+            print(f"File: {entity.name}")
+        elif isinstance(entity, Project):
+            print(f"Project: {entity.name}")
+        ```
+
+    Example: Retrieving entities by name
+        Get an entity by name and parent:
+
+        ```python
+        from synapseclient import Synapse
+        from synapseclient.models import get
```

Comment:
For discussion: I had always thought that models should contain the classes and the "services" and/or factory methods could live elsewhere. That said, I'm not opposed to this, just wanted to be mindful of the "models" module growing a lot